src/layers/basic.jl

```diff
@@ -226,3 +226,86 @@ end
 function Base.show(io::IO, b::SkipConnection)
   print(io, "SkipConnection(", b.layers, ", ", b.connection, ")")
 end
+
+"""
+    Bilinear(in1, in2, out)
+
+Creates a Bilinear layer, that operates over two inputs at the same time.
+Given parameters W and b, it's output is an interaction of the form 
+
+z_i = σ.(x * W_i * y' .+ b_i)
+
+All `z_i` are then concatenated across output dimensions `i` and samples.
+
+Bilinear can be used to create interactions between features as well.
+
+```
+d = Dense(10, 10)
+b = Bilinear(10, 10, 5)
+x = randn(Float32,10,9)
+
+#using Bilinear as the recombinator in a SkipConnection
+sc = SkipConnection(d, b)
+size(sc(x)) == (5,9)
+
+#using Bilinear in two data streams at once
+x = randn(Float32,10,9)
+y = randn(Float32,2,9)
+b = Bilinear(10, 2, 3)
+size(b(x,y)) == (3,9)
+
+#using Bilinear to generate interactions
+x = randn(Float32,11,7)
+b = Bilinear(11, 11, 3)
+size(b(x)) == (3,7)
+```
+"""
+struct Bilinear{A,B,S}
+  W::A
+  b::B
+  σ::S
+end
+
+@functor Bilinear
+
+Bilinear(W, b) = Bilinear(W, b, identity)
+
+function Bilinear(in1::Integer, in2::Integer, out::Integer, σ = identity;
+  initW = glorot_uniform, initb = zeros)
+  return Bilinear(initW(out, in1, in2), initb(out), σ)
+end
+
+function (a::Bilinear)(x::AbstractMatrix, y::AbstractMatrix)
+  if size(x, 2) ≠ size(y, 2)
+    error("Bilinear expected equal number of samples in both streams. Got $(size(x,2)) and $(size(y,2))")
+  end
+  W, b, σ = a.W, a.b, a.σ
+  Wp = permutedims(W, (2,1,3)) # could store in this order
+  o, i, j = size(W)
+  xs, ys = collecteachcol(x), collecteachcol(y)
+  Zs = map(xs, ys) do x,y
+    Wy = reshape(Wp, :,j) * y
+    xWy = vec(x' * reshape(Wy, j, o))
+  end
+  Z = reducehcat(Zs)
+  σ.(Z .+ b)
+end
+
+(a::Bilinear)(x::AbstractArray) = a(x,x)
+(a::Bilinear)(x::AbstractVector, y::AbstractVector) = vec(a(reshape(x, :,1), reshape(y, :,1)))
+
+function Base.show(io::IO, l::Bilinear)
+  print(io, "Bilinear(", size(l.W, 2), ", ", size(l.W, 3), ", ", size(l.W, 1))
+  l.σ == identity || print(io, ", ", l.σ)
+  print(io, ")")
+end
+
+collecteachcol(x::AbstractMatrix) = collect(eachcol(x))
```

Comment:
```suggestion
collecteachcol(x::AbstractMatrix) = [view(x, :,c) for c in axes(x,2)]
```
May actually be quicker without views, `x[:, c]`, but haven't checked.

Reply:
In the source code it's a Generator though , `eachcol(A::AbstractVecOrMat) = (view(A, :, i) for i in axes(A, 2))`

Reply:
Right, but I added `collect`. You could try without, I'm not sure whether the gradients still work correctly.

Reply:
Forgot to click the commit suggestion button, sorry for that, was testing locally